java/performance/src/test/java/org/apache/arrow/adapter/jdbc/JdbcAdapterBenchmarks.java

```diff
@@ -0,0 +1,331 @@
+  public static class RowConsumeState {
+
+    private Connection conn = null;
+
+    private ResultSet resultSet = null;
+
+    private BaseAllocator allocator;
+
+    private Statement statement;
+
+    private JdbcToArrowConfig config;
+
+    private ArrowVectorIterator iter;
+
+    private VectorSchemaRoot root;
+
+    @Setup(Level.Trial)
+    public void prepareState() throws Exception {
+      allocator = new RootAllocator(Integer.MAX_VALUE);
+      config = new JdbcToArrowConfigBuilder().setAllocator(allocator).setTargetBatchSize(1024).build();
+      Class.forName(DRIVER);
+      conn = DriverManager.getConnection(URL);
+
+      try (Statement stmt = conn.createStatement()) {
+        stmt.executeUpdate(CREATE_STATEMENT);
+      }
+
+      for (int i = 0; i < VALUE_COUNT; i++) {
+        // Insert data
+        try (PreparedStatement stmt = conn.prepareStatement(INSERT_STATEMENT)) {
+
+          stmt.setInt(1, i);
+          stmt.setLong(2, i);
+          stmt.setString(3, "test" + i);
+          stmt.setBoolean(4, i % 2 == 0);
+          stmt.executeUpdate();
+        }
+      }
+    }
+
+    @Setup(Level.Invocation)
+    public void prepareInvoke() throws Exception {
+      statement = conn.createStatement();
+      resultSet = statement.executeQuery(QUERY);
+
+      iter = JdbcToArrow.sqlToArrowVectorIterator(resultSet, config);
+      root = iter.next();
+      iter.compositeConsumer.resetVectorSchemaRoot(root);
+    }
+
+    @TearDown(Level.Invocation)
+    public void tearDownInvoke() throws Exception {
+      resultSet.close();
+      statement.close();
+      iter.close();
+    }
+
+    @TearDown(Level.Trial)
+    public void tearDownState() throws Exception {
+      try (Statement stmt = conn.createStatement()) {
+        stmt.executeUpdate(DROP_STATEMENT);
+      }
+      allocator.close();
+    }
+  }
+
+  /**
+   * Test {@link JdbcToArrow#sqlToArrowVectorIterator(ResultSet, JdbcToArrowConfig)}
+   * @return useless. To avoid DCE by JIT.
+   */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.MICROSECONDS)
+  public int testJdbcToArrow(JdbcState state) throws Exception {
+    int valueCount = 0;
+    try (ArrowVectorIterator iter = JdbcToArrow.sqlToArrowVectorIterator(state.resultSet, state.config)) {
+      while (iter.hasNext()) {
+        VectorSchemaRoot root = iter.next();
+        IntVector intVector = (IntVector) root.getFieldVectors().get(0);
+        valueCount += intVector.getValueCount();
+        root.close();
+      }
+    }
+    return valueCount;
+  }
+
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.MICROSECONDS)
+  public void consumeBenchmark(ConsumeState state) throws Exception {
+    state.intConsumer.resetValueVector(state.intVector);
+    state.longConsumer.resetValueVector(state.longVector);
+    for (int i = 0; i < VALUE_COUNT; i++) {
+      state.intConsumer.consume(state.resultSet);
+      state.longConsumer.consume(state.resultSet);
```

Comment:
does anything happen if you add a third consumer here?

Reply:
The performance improvement is still notable after adding a third consumer:

After:
JdbcAdapterBenchmarks.consumeBenchmark  avgt    5  167.243 ± 0.394  us/op

Before:
JdbcAdapterBenchmarks.consumeBenchmark  avgt    5  223.574 ± 1.791  us/op

Reply:
With 4 consumers (int, long, varchar, boolean), the performance improvement is still observable, but not as notable:

After
JdbcAdapterBenchmarks.consumeBenchmark  avgt    5  477.022 ± 20.044  us/op

Before
JdbcAdapterBenchmarks.consumeBenchmark  avgt    5  544.958 ± 20.120  us/op

This maybe caused by including a variable length field (varchar), which incurrs some extra cost that hides the performance improvement.

Reply:
It seems the above results confirm that the problem is caused by the loop in the CompositeConsumer?
To solve the problem, we need to create a CompositeConsumer by code generation?

Reply:
> It seems the above results confirm that the problem is caused by the loop in the CompositeConsumer?
To solve the problem, we need to create a CompositeConsumer by code generation?
Maybe.  One first step might be convert the composite consumer to use:
for (int x = 0; x < consumers.length; x++) {
  consumer[x].consume(...);
}
which is possibly more efficient.  Otherwise, yet it would seem that code generation might be necessary to see any real-world performance gains from this change.

Reply:
> Maybe. One first step might be convert the composite consumer to use:
> for (int x = 0; x < consumers.length; x++) {
> consumer[x].consume(...);
> }
> which is possibly more efficient. Otherwise, yet it would seem that code generation might be necessary to see any real-world performance gains from this change.

Thanks for the suggestion. I have tried the above method before, and it made almost no difference, though I agree that it really improves performance (I will change it accordingly). I think the most cost comes from the loop, which is even more expensive than a *if*. 

About code generation, I think it is only beneficial when the composite consumer is called repeated on a large volume of data, because dynamically compiling the generated code will take 10-100 ms.